### ISSUE TITLE
Create a BOSH release for etcd v3 (K8s will require v3 at some point in the future)

### ISSUE THREAD

**Issue by alejandrogoyen** [pivotal-cf-experimental/kubo-release#43]

**Comment by johnsonj** [comment]
# TL;DR

Looks good, the only failing scenarios aren't supported/required by Kubernetes. Scaling works fine. 

The next etcd thing we should do is investigate self-hosting. If that succeeds we should strip down etcd-release to support only what is needed for bootstrapping (no scaling, no support for multiple instances) or even start from scratch.

# Tricks for testing etcd-release:

- create a brand new bosh-lite
- rename `bosh-cli` (V2) to `bosh` and put it in the front of your path (alias is not enough) 
- upload stemcell
 bosh -e lite upload-stemcell https://bosh.io/d/stemcells/bosh-warden-boshlite-ubuntu-trusty-go_agent
- integration_config.json needs full URL of bosh-lite and CA cert:
```
echo '{
  "bosh":{
    "target": "https://192.168.50.4:25555/",
    "username": "admin",
    "password": "admin",
    "director_ca_cert": "~/dev/src/bosh-lite/ca/certs/ca.crt"
  }
}' > integration_config.json
```
- Go needs to be installed as a regular user, not as a system level package (I use GVM). You need write access to $GOROOT.
- Run the tests: 
```
bosh -e lite upload-stemcell https://bosh.io/d/stemcells/bosh-warden-boshlite-ubuntu-trusty-go_agent && EATS_CONFIG=$PWD/integration_config.json scripts/test
```

# Failures with etcdv3

The `turbulence` tests were not ran because they do not work on bosh-lite. This is a known issue in etcd-release.

## Failures
<details>
  <summary>
<li>[Fail] Migrate instance name change [It] migrates an etcd cluster sucessfully when instance name is changed /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/migrate_instance_name_change_test.go:98</li>
<li>[Fail] TLS rotation [BeforeEach] successfully rolls with new tls keys and cert  /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/tls_rotation_test.go:265</li>
<li>[Fail] Multiple instance rolling upgrade [It] persists data throughout the rolling upgrade /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/multiple_instance_rolling_upgrade_test.go:86</li>
  </summary>
  <code>

    [1495482442] deploy - 13/13 specs 
    ------------------------------
    • [SLOW TEST:607.073 seconds]
    TLS Upgrade keeps writing to an etcd cluster without interruption 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/tls_upgrade_test.go:248
    ------------------------------
    • [SLOW TEST:442.000 seconds]
    Scaling down instances with TLS scales from 5 to 3 nodes 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/scale_down_instances_test.go:92
    ------------------------------
    STEP: setting a persistent value
    STEP: deploying with a new name
    • Failure [260.684 seconds]
    Migrate instance name change [It] migrates an etcd cluster sucessfully when instance name is changed 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/migrate_instance_name_change_test.go:149

      Timed out after 60.001s.
      Expected
          <[]bosh.VM | len:4, cap:4>: [
              {ID: "", Index: 0, State: "failing", JobName: "new_etcd", IPs: nil},
              {
                  ID: "",
                  Index: 0,
                  State: "running",
                  JobName: "testconsumer",
                  IPs: nil,
              },
              {ID: "", Index: 2, State: "failing", JobName: "new_etcd", IPs: nil},
              {ID: "", Index: 1, State: "running", JobName: "new_etcd", IPs: nil},
          ]
      to consist of
          <[]interface {} | len:1, cap:1>: [
              [
                  {ID: "", Index: 0, State: "running", JobName: "new_etcd", IPs: nil},
                  {ID: "", Index: 1, State: "running", JobName: "new_etcd", IPs: nil},
                  {ID: "", Index: 2, State: "running", JobName: "new_etcd", IPs: nil},
                  {
                      ID: "",
                      Index: 0,
                      State: "running",
                      JobName: "testconsumer",
                      IPs: nil,
                  },
              ],
          ]

      /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/migrate_instance_name_change_test.go:98
    ------------------------------
    • [SLOW TEST:341.454 seconds]
    consistency checker with TLS checks if etcd consistency checker reports failures to bosh during split brain 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/consistency_checker_test.go:125
    ------------------------------
    • Failure in Spec Setup (BeforeEach) [130.603 seconds]
    TLS rotation [BeforeEach] successfully rolls with new tls keys and cert 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/tls_rotation_test.go:430

      Expected error:
          <bosh.TaskError>: {
              Code: 450002,
              Message: "Timed out sending 'prepare' to 38edbd9e-a27f-44f3-892d-ef310439cf82 after 45 seconds",
          }
          task error: 450002 has occurred: Timed out sending 'prepare' to 38edbd9e-a27f-44f3-892d-ef310439cf82 after 45 seconds
      not to have occurred

      /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/tls_rotation_test.go:265
    ------------------------------
    • [SLOW TEST:321.866 seconds]
    Multiple instance rolling deploys without TLS persists data throughout the rolling deploy 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/multiple_instance_rolling_deploy_test.go:140
    ------------------------------
    • [SLOW TEST:336.013 seconds]
    Scaling up instances without TLS scales from 3 to 5 nodes 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/scale_up_instances_test.go:92
    ------------------------------
    • [SLOW TEST:316.028 seconds]
    Scaling down instances without TLS scales from 5 to 3 nodes 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/scale_down_instances_test.go:92
    ------------------------------
    • [SLOW TEST:311.182 seconds]
    consistency checker without TLS checks if etcd consistency checker reports failures to bosh during split brain 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/consistency_checker_test.go:125
    ------------------------------
    STEP: setting a persistent value
    STEP: deploying the latest dev build of etcd-release
    • Failure [416.361 seconds]
    Multiple instance rolling upgrade [It] persists data throughout the rolling upgrade 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/multiple_instance_rolling_upgrade_test.go:142

      Expected error:
          <bosh.TaskError>: {
              Code: 400007,
              Message: "'etcd/0 (7f69b196-d601-421e-b9f5-11038a8c17b1)' is not running after update. Review logs for failed jobs: etcd",
          }
          task error: 400007 has occurred: 'etcd/0 (7f69b196-d601-421e-b9f5-11038a8c17b1)' is not running after update. Review logs for failed jobs: etcd
      not to have occurred

      /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/multiple_instance_rolling_upgrade_test.go:86
    ------------------------------
    • [SLOW TEST:338.679 seconds]
    Multiple instance rolling deploys with TLS persists data throughout the rolling deploy 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/multiple_instance_rolling_deploy_test.go:140
    ------------------------------
    • [SLOW TEST:501.974 seconds]
    split brain one leader bubbles up a failure from consistency checker to bosh 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/split_brain_one_leader_test.go:205
    ------------------------------
    • [SLOW TEST:421.918 seconds]
    Scaling up instances with TLS scales from 3 to 5 nodes 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/scale_up_instances_test.go:92
    ------------------------------


    Summarizing 3 Failures:

    [Fail] Migrate instance name change [It] migrates an etcd cluster sucessfully when instance name is changed 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/migrate_instance_name_change_test.go:98

    [Fail] TLS rotation [BeforeEach] successfully rolls with new tls keys and cert 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/tls_rotation_test.go:265

    [Fail] Multiple instance rolling upgrade [It] persists data throughout the rolling upgrade 
    /usr/local/google/home/jrjohnson/dev/src/github.com/cloudfoundry-incubator/etcd-release/src/acceptance-tests/deploy/multiple_instance_rolling_upgrade_test.go:86
  </code>
</details>

**Comment by johnsonj** [comment]
Release: https://storage.googleapis.com/kubo-etcd/kubo-etcd-release.108%2Bdev.2.tgz

**Comment by alex-slynko** [comment]
work continued in #53
